File: web/scripts/prepare-archive-assets.ts
import { mkdir, readdir, readFile, rm, stat, writeFile } from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import {
	CanvasRenderingContext2D,
	DOMMatrix,
	Image,
	Path2D,
	createCanvas,
} from "@napi-rs/canvas";
import pdf from "pdf-parse";
import * as pdfjsLib from "pdfjs-dist/legacy/build/pdf.mjs";

type ArchiveRecord = {
	title: string;
	slug: string;
	date: string;
	volumeName: string;
	regionalVariant: string;
	sourcePdf: string;
	coverImage: string;
	pageImages: string[];
	extractedText: string;
	draftAbstract: string;
};

type TextItem = {
	str?: string;
	transform?: number[];
	height?: number;
	width?: number;
	hasEOL?: boolean;
};

const DEFAULT_INPUT_DIR =
	"C:/Users/jonbr/source/repos/freedom-times-content/original-freedom-times-pdf/split";
const DEFAULT_OUTPUT_DIR = ".generated/archive-import";
const ABSTRACT_MAX_LENGTH = 320;

Object.assign(globalThis, {
	Image,
	Path2D,
	DOMMatrix,
	CanvasRenderingContext2D,
});

function getArgValue(flag: string): string | undefined {
	const index = process.argv.indexOf(flag);
	if (index === -1) return undefined;
	return process.argv[index + 1];
}

function slugToTitle(slug: string): string {
	const parts = slug.split("-");
	if (parts.length >= 4 && /^\d{4}$/.test(parts[0] ?? "")) {
		const date = parts.slice(0, 3).join("-");
		const rest = parts.slice(3).map(capitalizeWord).join(" ");
		return `${date} ${rest}`;
	}
	return parts.map(capitalizeWord).join(" ");
}

function capitalizeWord(word: string): string {
	if (!word) return word;
	if (word.toUpperCase() === word) return word;
	return word.charAt(0).toUpperCase() + word.slice(1);
}

function extractDateFromName(filename: string): string {
	const match = filename.match(/^(\d{4}-\d{2}-\d{2})/);
	if (!match) {
		throw new Error(`Could not extract date from filename: ${filename}`);
	}
	return `${match[1]}T00:00:00.000Z`;
}

function sanitizeSegment(value: string): string {
	return value
		.toLowerCase()
		.replace(/[^a-z0-9]+/g, '-')
		.replace(/^-+|-+$/g, '')
		|| 'unknown';
}

function deriveEditionMetadata(slug: string): { volumeName: string; regionalVariant: string } {
	const tokens = slug.split('-');
	const bodyTokens = tokens.slice(3);
	const freedomTimesIndex = bodyTokens.findIndex(
		(token, index) => token === 'freedom' && bodyTokens[index + 1] === 'times',
	);
	const prefix = freedomTimesIndex >= 0 ? bodyTokens.slice(0, freedomTimesIndex) : bodyTokens;

	let volumeName = 'main-edition';
	let regionalVariant = 'default';

	const volumeIndex = prefix.findIndex((token) => /^vol(?:ume)?(?:-?\d+)?$/i.test(token));
	if (volumeIndex >= 0) {
		const part = prefix[volumeIndex + 1];
		volumeName = part ? `${prefix[volumeIndex]}-${part}` : prefix[volumeIndex];
	}

	const regionalTokens = prefix.filter((_, index) => index !== volumeIndex && index !== volumeIndex + 1);
	if (regionalTokens.length > 0) {
		regionalVariant = regionalTokens.join('-');
	}

	return {
		volumeName: sanitizeSegment(volumeName),
		regionalVariant: sanitizeSegment(regionalVariant),
	};
}

function sanitizeText(text: string): string {
	return text
		.replace(/\r/g, "\n")
		.replace(/\u0000/g, " ")
		.replace(/[ \t]+/g, " ")
		.replace(/\n{3,}/g, "\n\n")
		.trim();
}

function buildDraftAbstract(text: string, filename: string): string {
	const lines = sanitizeText(text)
		.split(/\n+/)
		.map((line) => line.trim())
		.filter((line) => line.length > 30)
		.filter((line) => !/^page\s+\d+$/i.test(line))
		.filter((line) => !/^freedom times$/i.test(line));

	const joined = lines.join(" ").replace(/\s+/g, " ").trim();
	if (!joined) {
		return `Archive issue for ${extractDateFromName(filename).slice(0, 10)}.`;
	}

	const sentenceMatch = joined.match(/(.{80,320}?[.!?])(?=\s|$)/);
	const candidate = sentenceMatch?.[1] ?? joined.slice(0, ABSTRACT_MAX_LENGTH);
	return candidate.trim();
}

async function renderAllPages(
	pdfBuffer: Buffer,
	outputDir: string,
	baseName: string,
): Promise<{ coverImage: string; pageImages: string[] }> {
	const loadingTask = pdfjsLib.getDocument({ data: new Uint8Array(pdfBuffer) });
	const document = await loadingTask.promise;
	const pageImages: string[] = [];

	for (let pageNumber = 1; pageNumber <= document.numPages; pageNumber += 1) {
		const page = await document.getPage(pageNumber);
		const viewport = page.getViewport({ scale: 2 });
		const canvas = createCanvas(Math.ceil(viewport.width), Math.ceil(viewport.height));
		const context = canvas.getContext('2d');
		const filename = `${baseName}--page-${String(pageNumber).padStart(3, '0')}.png`;
		const outputPath = path.join(outputDir, filename);

		await page.render({
			canvasContext: context,
			viewport,
		}).promise;

		await writeFile(outputPath, canvas.toBuffer('image/png'));
		await page.cleanup();
		pageImages.push(outputPath);
	}

	await document.destroy();

	if (pageImages.length === 0) {
		throw new Error('PDF has no pages to render');
	}

	return {
		coverImage: pageImages[0],
		pageImages,
	};
}

async function extractPageOneText(pdfBuffer: Buffer): Promise<string> {
	const loadingTask = pdfjsLib.getDocument({ data: new Uint8Array(pdfBuffer) });
	const document = await loadingTask.promise;
	const page = await document.getPage(1);
	const content = await page.getTextContent();
	const items = (content.items as TextItem[])
		.map((item) => item.str?.trim() ?? "")
		.filter(Boolean);
	await page.cleanup();
	await document.destroy();
	return items.join("\n");
}

async function processPdf(inputDir: string, outputDir: string, fileName: string): Promise<ArchiveRecord> {
	const sourcePdf = path.join(inputDir, fileName);
	const slug = fileName.replace(/\.pdf$/i, "");
	const date = extractDateFromName(fileName);
	const title = slugToTitle(slug);
	const { volumeName, regionalVariant } = deriveEditionMetadata(slug);
	const recordDir = path.join(outputDir, slug);
	const extractedText = path.join(recordDir, "text.txt");
	const imageBaseName = `${slug}--vol-${volumeName}--region-${regionalVariant}`;

	await mkdir(recordDir, { recursive: true });
	const pdfBuffer = await readFile(sourcePdf);
	const parsed = await pdf(pdfBuffer);
	const firstPageText = await extractPageOneText(pdfBuffer);
	const mergedText = sanitizeText(`${firstPageText}\n\n${parsed.text}`);
	const draftAbstract = buildDraftAbstract(mergedText, fileName);
	const { coverImage, pageImages } = await renderAllPages(pdfBuffer, recordDir, imageBaseName);

	await writeFile(extractedText, mergedText, "utf8");

	return {
		title,
		slug,
		date,
		volumeName,
		regionalVariant,
		sourcePdf,
		coverImage,
		pageImages,
		extractedText,
		draftAbstract,
	};
}

async function main(): Promise<void> {
	const inputDir = path.resolve(getArgValue("--input") ?? DEFAULT_INPUT_DIR);
	const outputDir = path.resolve(getArgValue("--output") ?? DEFAULT_OUTPUT_DIR);

	const inputStat = await stat(inputDir);
	if (!inputStat.isDirectory()) {
		throw new Error(`Input path is not a directory: ${inputDir}`);
	}

	await rm(outputDir, { recursive: true, force: true });
	await mkdir(outputDir, { recursive: true });

	const fileNames = (await readdir(inputDir))
		.filter((entry) => entry.toLowerCase().endsWith(".pdf"))
		.sort((left, right) => left.localeCompare(right));

	if (fileNames.length === 0) {
		throw new Error(`No PDFs found in ${inputDir}`);
	}

	const records: ArchiveRecord[] = [];
	for (const fileName of fileNames) {
		records.push(await processPdf(inputDir, outputDir, fileName));
		console.log(`Prepared ${fileName}`);
	}

	await writeFile(path.join(outputDir, "manifest.json"), JSON.stringify(records, null, 2), "utf8");
	console.log(`Prepared ${records.length} archive PDFs into ${outputDir}`);
}

main().catch((error: unknown) => {
	console.error(error instanceof Error ? error.message : error);
	process.exitCode = 1;
});

File: web/scripts/prepare_archive_assets.py
import argparse
import json
import re
import shutil
from pathlib import Path

import fitz

DEFAULT_INPUT_DIR = Path(r"C:\Users\jonbr\source\repos\freedom-times-content\original-freedom-times-pdf\split")
DEFAULT_OUTPUT_DIR = Path(".generated/archive-import")
DEFAULT_COVERS_DIR = "covers"
DEFAULT_TEXT_DIR = "text"
ABSTRACT_MIN = 80
ABSTRACT_MAX = 320


def slug_to_title(slug: str) -> str:
    parts = slug.split("-")
    return " ".join(part.upper() if part.isupper() else part.capitalize() for part in parts)


def date_from_filename(filename: str) -> str:
    match = re.match(r"^(\d{4}-\d{2}-\d{2})", filename)
    if not match:
        raise ValueError(f"Could not extract date from filename: {filename}")
    return f"{match.group(1)}T00:00:00.000Z"


def sanitize_segment(value: str) -> str:
    cleaned = re.sub(r"[^a-z0-9]+", "-", value.lower()).strip("-")
    return cleaned or "unknown"


def derive_edition_metadata(slug: str) -> tuple[str, str]:
    tokens = slug.split("-")
    body_tokens = tokens[3:]

    freedom_times_idx = -1
    for idx in range(len(body_tokens) - 1):
        if body_tokens[idx] == "freedom" and body_tokens[idx + 1] == "times":
            freedom_times_idx = idx
            break

    prefix = body_tokens[:freedom_times_idx] if freedom_times_idx >= 0 else body_tokens
    volume_name = "main-edition"
    regional_variant = "default"

    volume_idx = -1
    for idx, token in enumerate(prefix):
        if re.match(r"^vol(?:ume)?(?:-?\d+)?$", token, flags=re.IGNORECASE):
            volume_idx = idx
            break

    if volume_idx >= 0:
        second = prefix[volume_idx + 1] if volume_idx + 1 < len(prefix) else ""
        volume_name = f"{prefix[volume_idx]}-{second}" if second else prefix[volume_idx]

    regional_tokens = [
        token
        for idx, token in enumerate(prefix)
        if idx not in (volume_idx, volume_idx + 1)
    ]
    if regional_tokens:
        regional_variant = "-".join(regional_tokens)

    return sanitize_segment(volume_name), sanitize_segment(regional_variant)


def sanitize_text(value: str) -> str:
    value = value.replace("\r", "\n").replace("\x00", " ")
    value = re.sub(r"[ \t]+", " ", value)
    value = re.sub(r"\n{3,}", "\n\n", value)
    return value.strip()


def build_draft_abstract(text: str, fallback_date: str) -> str:
    lines = [line.strip() for line in sanitize_text(text).splitlines()]
    lines = [line for line in lines if len(line) > 30]
    lines = [line for line in lines if not re.match(r"^page\s+\d+$", line, flags=re.IGNORECASE)]
    lines = [line for line in lines if line.lower() != "freedom times"]

    joined = re.sub(r"\s+", " ", " ".join(lines)).strip()
    if not joined:
        return f"Archive issue for {fallback_date}."

    sentence_match = re.search(r"(.{80,320}?[.!?])(?=\s|$)", joined)
    candidate = sentence_match.group(1) if sentence_match else joined[:ABSTRACT_MAX]
    return candidate.strip()


def extract_pdf_text(document: fitz.Document) -> str:
    chunks: list[str] = []
    for page in document:
        chunks.append(page.get_text("text"))
    return sanitize_text("\n\n".join(chunks))


def render_all_pages(document: fitz.Document, output_dir: Path, base_name: str) -> list[Path]:
    page_paths: list[Path] = []
    for idx, page in enumerate(document, start=1):
        filename = f"{base_name}--page-{idx:03d}.png"
        output_path = output_dir / filename
        pix = page.get_pixmap(matrix=fitz.Matrix(2, 2), alpha=False)
        pix.save(output_path)
        page_paths.append(output_path)
    return page_paths


def process_pdf(pdf_path: Path, covers_dir: Path, text_dir: Path) -> dict:
    slug = pdf_path.stem
    date_iso = date_from_filename(pdf_path.name)
    date_only = date_iso[:10]
    title = slug_to_title(slug)
    volume_name, regional_variant = derive_edition_metadata(slug)
    image_base_name = f"{slug}--vol-{volume_name}--region-{regional_variant}"

    cover_path = covers_dir / f"{image_base_name}--page-001.png"
    text_path = text_dir / f"{slug}.txt"

    with fitz.open(pdf_path) as document:
        text = extract_pdf_text(document)
        page_paths = render_all_pages(document, covers_dir, image_base_name)

    text_path.write_text(text, encoding="utf-8")

    if not page_paths:
        raise ValueError(f"PDF has no pages to render: {pdf_path}")

    return {
        "title": title,
        "slug": slug,
        "date": date_iso,
        "volumeName": volume_name,
        "regionalVariant": regional_variant,
        "sourcePdf": str(pdf_path),
        "coverImage": str(cover_path),
        "pageImages": [str(path) for path in page_paths],
        "extractedText": str(text_path),
        "draftAbstract": build_draft_abstract(text, date_only),
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--input", default=str(DEFAULT_INPUT_DIR))
    parser.add_argument("--output", default=str(DEFAULT_OUTPUT_DIR))
    args = parser.parse_args()

    input_dir = Path(args.input).resolve()
    output_dir = Path(args.output).resolve()
    covers_dir = output_dir / DEFAULT_COVERS_DIR
    text_dir = output_dir / DEFAULT_TEXT_DIR

    if not input_dir.is_dir():
        raise ValueError(f"Input path is not a directory: {input_dir}")

    if output_dir.exists():
        shutil.rmtree(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)
    covers_dir.mkdir(parents=True, exist_ok=True)
    text_dir.mkdir(parents=True, exist_ok=True)

    pdf_files = sorted(input_dir.glob("*.pdf"))
    if not pdf_files:
        raise ValueError(f"No PDFs found in {input_dir}")

    records = []
    for pdf_file in pdf_files:
        record = process_pdf(pdf_file, covers_dir, text_dir)
        records.append(record)
        print(f"Prepared {pdf_file.name}")

    manifest_path = output_dir / "manifest.json"
    manifest_path.write_text(json.dumps(records, indent=2), encoding="utf-8")
    print(f"Prepared {len(records)} archive PDFs into {output_dir}")


if __name__ == "__main__":
    main()
